### PR TITLE
The order of raw1.info["bads"] should not matter when concatenating with raw0.info["bads"] #11501

### LINKED ISSUE #11501: The order of raw1.info["bads"] should not matter when concatenating with raw0.info["bads"]
### Description of the problem

I have a list of `raws` and combine them using `raw_concat = concatenate_raws(raws)`. However, I got the following error:

```Python
Exception has occurred: ValueError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
raw[1].info[bads] must match:
['LFP_R_4_STN_BS_RSTN', 'LFP_R_4_STN_BS'] != ['LFP_R_4_STN_BS', 'LFP_R_4_STN_BS_RSTN']
  File "/Users/moritzgerster/Library/CloudStorage/Dropbox/Code/BIDS_Neumann_ECoG_LFP/scripts/utils.py", line 482, in _combine_raws
    raw_concat = concatenate_raws(raws)
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "<decorator-gen-238>", line 12, in concatenate_raws
  File "/Users/moritzgerster/anaconda3/envs/bids_neu/lib/python3.11/site-packages/mne/io/base.py", line 2542, in concatenate_raws
    _ensure_infos_match(info1=raws[0].info, info2=raw.info,
  File "/Users/moritzgerster/anaconda3/envs/bids_neu/lib/python3.11/site-packages/mne/io/meas_info.py", line 2884, in _ensure_infos_match
    raise ValueError(f'{name}.info[\'nchan\'] must match')
ValueError: raws[1].info['nchan'] must match

During handling of the above exception, another exception occurred:

  File "/Users/moritzgerster/anaconda3/envs/bids_neu/lib/python3.11/site-packages/mne/io/base.py", line 2495, in _check_raw_compatibility
    raise ValueError(
  File "/Users/moritzgerster/anaconda3/envs/bids_neu/lib/python3.11/site-packages/mne/io/base.py", line 1641, in append
    _check_raw_compatibility(all_raws)
  File "/Users/moritzgerster/anaconda3/envs/bids_neu/lib/python3.11/site-packages/mne/io/base.py", line 2551, in concatenate_raws
    raws[0].append(raws[1:], preload)
  File "<decorator-gen-238>", line 12, in concatenate_raws
  File "/Users/moritzgerster/Library/CloudStorage/Dropbox/Code/BIDS_Neumann_ECoG_LFP/scripts/utils.py", line 492, in _combine_raws
    raw_concat = concatenate_raws(raws)
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/moritzgerster/Library/CloudStorage/Dropbox/Code/BIDS_Neumann_ECoG_LFP/scripts/_02_fooof_slopes.py", line 59, in get_fitted_spectra
    raw_concat = _combine_raws(bids_paths, save_path=bids_path)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/moritzgerster/Library/CloudStorage/Dropbox/Code/BIDS_Neumann_ECoG_LFP/scripts/_02_fooof_slopes.py", line 36, in psd_fooof_to_dataframe
    dics = get_fitted_spectra(paths_to_sub_in_cond)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/moritzgerster/Library/CloudStorage/Dropbox/Code/BIDS_Neumann_ECoG_LFP/scripts/main.py", line 28, in <module>
    psd_fooof_to_dataframe()
  File "/Users/moritzgerster/anaconda3/envs/bids_neu/lib/python3.11/runpy.py", line 88, in _run_code
    exec(code, run_globals)
  File "/Users/moritzgerster/anaconda3/envs/bids_neu/lib/python3.11/runpy.py", line 198, in _run_module_as_main (Current frame)
    return _run_code(code, main_globals, None,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: raw[1].info[bads] must match:
['LFP_R_4_STN_BS_RSTN', 'LFP_R_4_STN_BS'] != ['LFP_R_4_STN_BS', 'LFP_R_4_STN_BS_RSTN']
```

As you can see the bad channels 
`['LFP_R_4_STN_BS_RSTN', 'LFP_R_4_STN_BS'] != ['LFP_R_4_STN_BS', 'LFP_R_4_STN_BS_RSTN']`
are identical in both cases, only their order is wrong.

https://github.com/mne-tools/mne-python/blob/603a96da58ddba3c2c05d3a01639abd669cc9543/mne/io/base.py#L2506-L2513

The above lines of code should be replaced with:

```Python
        for key in ('nchan', 'sfreq'):
            a, b = raw[ri].info[key], raw[0].info[key]
            if a != b:
                raise ValueError(
                    f'raw[{ri}].info[{key}] must match:\n'
                    f'{repr(a)} != {repr(b)}')
        if not set(raw[ri].info['bads']) == set(raw[0].info['bads']):
            raise ValueError('raw[%d][\'info\'][\'bads\'] must match' % ri)
        if not set(raw[ri].info['ch_names']) == set(raw[0].info['ch_names']):
            raise ValueError('raw[%d][\'info\'][\'ch_names\'] must match' % ri)
```

### Steps to reproduce

```Python
import mne
from mne import create_info
from mne.io import RawArray
from numpy.random import default_rng

def create_toy_data(n_channels=3, sfreq=250, seed=None):
    rng = default_rng(seed)
    data = rng.standard_normal(size=(n_channels, 50 * sfreq)) * 5e-6
    info = create_info(n_channels, sfreq, "eeg")
    return RawArray(data, info)

raw0 = create_toy_data()
raw1 = create_toy_data()

raw0.info["bads"] = ["0", "1"]
raw1.info["bads"] = ["1", "0"]

raw_concat = mne.concatenate_raws([raw0, raw1])
```

### Link to data

_No response_

### Expected results

Concatenation should work.

### Actual results

ValueError.

### Additional information

```Python
Platform:         macOS-13.2.1-arm64-arm-64bit
Python:           3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:26:40) [Clang 14.0.6 ]
Executable:       /Users/moritzgerster/anaconda3/envs/bids_neu/bin/python
CPU:              arm: 10 cores
Memory:           16.0 GB

mne:              1.3.0
numpy:            1.23.4 {OpenBLAS 0.3.21 with 10 threads}
scipy:            1.9.3
matplotlib:       3.6.2 {backend=MacOSX}

sklearn:          1.1.3
numba:            Not found
nibabel:          5.0.1
nilearn:          Not found
dipy:             Not found
openmeeg:         Not found
cupy:             Not found
pandas:           1.5.1
pyvista:          Not found
pyvistaqt:        Not found
ipyvtklink:       Not found
vtk:              Not found
qtpy:             Not found
ipympl:           Not found
pyqtgraph:        Not found
pooch:            v1.6.0

mne_bids:         0.12
mne_nirs:         Not found
mne_features:     Not found
mne_qt_browser:   Not found
mne_connectivity: Not found
mne_icalabel:     Not found
```

### PR DESCRIPTION
**Pull request by moritz-gerster**
#### Reference issue
Fixes #11501.

#### What does this implement/fix?
The order of bad channels no longer matters.

Note, that this is already correctly implemented in `_ensure_infos_match`:
https://github.com/mne-tools/mne-python/blob/603a96da58ddba3c2c05d3a01639abd669cc9543/mne/io/meas_info.py#L2903-L2904

It's only wrong in `_check_raw_compatibility(raw)`:
https://github.com/mne-tools/mne-python/blob/603a96da58ddba3c2c05d3a01639abd669cc9543/mne/io/base.py#L2501-L2513

### CODE REVIEW

File: mne/io/base.py
Comment (larsoner) [inline mne/io/base.py:2512]: I'm surprised no style checker complained about this, so I'll have to do it :)
```suggestion
        if set(raw[ri].info['bads']) != set(raw[0].info['bads']):
```
A little BUG entry in `doc/changes/latest.inc` would be good, too!

File: mne/io/base.py
Comment (larsoner) [inline mne/io/base.py:2513]: ... also this line is less informative than the one above because it doesn't give the differences. You could use something like `sorted(set(a).symmetric_difference(set(b)))` to get the entries that don't match and print them like the line above

**Comment by moritz-gerster** [comment]
Shit I got confused by git. I hope I didn't mess everything up. I think some files were moved to the cloud during a commit, that might be why there are so many file changes

**Comment by larsoner** [comment]
Yikes, a ton of files deleted :)

I looked through your commits and saw which one did this (bff61e8c0bb03b9385b7ebad548f66023cb650a9), and then did:
```
git revert bff61e8c0bb03b9385b7ebad548f66023cb650a9
git checkout bff61e8c0bb03b9385b7ebad548f66023cb650a9 mne/io/base.py
git commit -am 'FIX: Try this'
git push
```
Looks like the diff is now `+24 −13` so probably closer to correct at least. Locally if you `git pull origin compare_bads` you should get to the same state, then you can fix whatever is still broken/missing!

File: mne/io/base.py
Comment (larsoner) [inline mne/io/base.py:2503]: please put back as `ri`. We try to avoid bare `i` (hard to read, get confused with `1j`, etc.) and `ri` is even a tiny bit more informative as it makes it clear it's the "raw iterator" rather than just "iterator".

File: mne/io/base.py
Comment (larsoner) [inline mne/io/base.py:2520]: This code also could be more DRY -- it should exist in a loop like
```
for kind in ('bads', 'ch_names'):
```
... but this also now does a symmetric diff on the channel names, which I don't think we've discussed yet. I actually don't think this will work -- I'd expect the user to need to do a minual `reorder_channels` before being able to concat successfully/correctly. So for `ch_names` the order does actually matter...

**Comment by moritz-gerster** [comment]
> Looks like the diff is now `+24 −13` so probably closer to correct at least. Locally if you `git pull origin compare_bads` you should get to the same state, then you can fix whatever is still broken/missing!

Thank you for your help! ☺️

**Comment by moritz-gerster** [comment]
Dear @larsoner, are the changes good to go now? :)

**Review by agramfort — COMMENTED** [review]
diff looks good @moritz-gerster 

any chance you can update or add a test for this?

**Comment by moritz-gerster** [comment]
> any chance you can update or add a test for this?

@agramfort: Check ✅

**Comment by moritz-gerster** [comment]
> Can you move your testing there to be consistent? Also rename your test test_concatenate_raws_bads_order

@agramfort done :)

File: mne/io/base.py
Comment (larsoner) [reply mne/io/base.py]: @moritz-gerster I think this comment has not been addressed. I do not think you should use `set` with channels like you do (now) for bads, so I think this `chs` part should be reverted / go back to old behavior, ideally with a test that fails if the channel *orders* do not match (it should IIRC!)

**Review by larsoner — CHANGES_REQUESTED** [review]
Marking as requesting changes so we don't forget to undo the chs-as-set check

**Review by agramfort — APPROVED** [review]

**Comment by moritz-gerster** [comment]
@larsoner is there more I should do? :)

**Comment by agramfort** [comment]
@moritz-gerster you need to fix conflicts

**Comment by moritz-gerster** [comment]
> @moritz-gerster you need to fix conflicts

@agramfort  done ✅

**Review by agramfort — APPROVED** [review]
@larsoner I let you merge

File: mne/io/base.py
Comment (larsoner) [reply mne/io/base.py]: @moritz-gerster I think this comment has not been addressed. I'll un-resolve it.

I noticed on this read through that it's a `set` check on `main` so you can ignore that part -- but I think it could still be more DRY by using a loop, right?

**Comment by moritz-gerster** [comment]
@larsoner is it good now?

**Comment by moritz-gerster** [comment]
@larsoner good to go now?

**Comment by agramfort** [comment]
@moritz-gerster lets wait for main to be green and lets merge this after its green again

**Review by larsoner — APPROVED** [review]

**Comment by larsoner** [comment]
Thanks @moritz-gerster !

**Pull request merged**
